cli/lsp/performance.rs

```diff
@@ -0,0 +1,165 @@
+// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
+
+use deno_core::serde::Deserialize;
+use deno_core::serde::Serialize;
+use std::collections::HashMap;
+use std::collections::VecDeque;
+use std::sync::Arc;
+use std::sync::Mutex;
+use std::time::Duration;
+use std::time::Instant;
+
+#[derive(Debug, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct PerformanceAverage {
+  pub name: String,
+  pub count: u32,
+  pub average_duration: u32,
+}
+
+/// A structure which serves as a start of a measurement span.
+#[derive(Debug)]
+pub struct PerformanceMark {
+  name: String,
+  count: u32,
+  start: Instant,
+}
+
+/// A structure which holds the information about the measured span.
+#[derive(Debug, Clone)]
+pub struct PerformanceMeasure {
+  pub name: String,
+  pub count: u32,
+  pub duration: Duration,
+}
+
+impl From<PerformanceMark> for PerformanceMeasure {
+  fn from(value: PerformanceMark) -> Self {
+    Self {
+      name: value.name,
+      count: value.count,
+      duration: value.start.elapsed(),
+    }
+  }
+}
+
+/// A simple structure for marking a start of something to measure the duration
+/// of and measuring that duration.  Each measurement is identified by a string
+/// name and a counter is incremented each time a new measurement is marked.
+///
+/// The structure will limit the size of measurements to the most recent 1000,
+/// and will roll off when that limit is reached.
+#[derive(Debug, Clone)]
```

Comment:
Does it need to be Clone? This will duplicate the size value...

Reply:
yes, because `LanguageServer` is `Clone` and all of its properties therefore need to be `Clone`.

Reply:
but `LanguageServer` appears not to require clone, so I will remove both.